server/src/server/kb/concept/TypeConverter.java

```
@@ -0,0 +1,133 @@
+            .put(AttributeType.DataType.BOOLEAN, new Default<Boolean>())
+            .put(AttributeType.DataType.DATE, new DateConverter())
+            .put(AttributeType.DataType.DOUBLE, new DoubleConverter())
+            .put(AttributeType.DataType.FLOAT, new FloatConverter())
+            .put(AttributeType.DataType.INTEGER, new IntegerConverter())
+            .put(AttributeType.DataType.LONG, new LongConverter())
+            .put(AttributeType.DataType.STRING, new Default<String>())
+            .build();
+
+
+    public static <DESERIALISED> TypeConverter<DESERIALISED> of(AttributeType.DataType<DESERIALISED> dataType) {
+        TypeConverter<?> converter = converters.get(dataType);
+        if (converter == null){
+            throw new UnsupportedOperationException("Unsupported DataType: " + dataType.toString());
+        }
+        return (TypeConverter<DESERIALISED>) converter;
+
+    }
+
+    public static class Default<DESERIALISED> extends TypeConverter<DESERIALISED>{
+        @Override
+        public Object serialise(DESERIALISED value) { return value; }
+
+        @Override
+        public DESERIALISED deserialise(Object value) { return (DESERIALISED) value;}
+    }
+
+    public static class DateConverter extends TypeConverter<LocalDateTime> {
+        @Override
+        public Object serialise(LocalDateTime value) {
+            return value.atZone(ZoneId.of("Z")).toInstant().toEpochMilli();
+        }
+
+        @Override
+        public LocalDateTime deserialise(Object value) {
+            if (value instanceof LocalDateTime) return (LocalDateTime) value;
+            if (value instanceof Long) return LocalDateTime.ofInstant(Instant.ofEpochMilli((Long) value), ZoneId.of("Z"));
+            throw new ClassCastException();
+        }
+    }
+
+    public static class DoubleConverter extends TypeConverter<Double> {
+        @Override
+        public Double deserialise(Object value) {
```

I think the input arguments here should definitely be restricted to `Number`? Along with every other `TypeConverter`, for that matter.

---

We still want to throw if it is an unsupported conversion/not a Number, so I don't think so. Also that way the whole conversion logic is in one place.

---

Ok, i changed my mind!